### ISSUE TITLE
Missing Homematic Device RSSI value displayed as -65.535 dBm

### ISSUE THREAD

**Issue by marc965** [SukramJ/hahomematic#178]
**custom_component/hahomematic version (if applicable):**
0.21.1

**Home Assistant version (if applicable):**
2021.12.9

**CCU version:**
- Model (CCU2/3, Raspi3/4, NUC, ProxmoxVE, ...): CCU2/3
- OS (Original, RaspberryMatic, debmatic/...): RaspberryMatic
- Version: 3.61.7.20211218

**Problem-relevant configuration:**

Do you use tls? no
Do you use callback? no
Do you use username and password? yes
Which interfaces do you use? BidCos-RF, HmIP-RF

**Describe the bug**
If there is no RSSI value transmitted from RaspberryMatic, HA displayed -65.535 dBm. Sometimes the negative sign of this value is omitted or not transfered properly.
This behavior only observed for Homematic devices, Homematic IP devices are displayed and refreshed properly.

**Screenshots**
Homematic Manager:
![](https://user-images.githubusercontent.com/85791333/149384711-a7f394a8-7ea6-4c59-91db-01ffdd9fde3d.png)
RaspberryMatic WebUI:
![](https://user-images.githubusercontent.com/85791333/149385450-b2511f48-72d9-4285-998f-723751e8f024.png)
Homeassistant:
![](https://user-images.githubusercontent.com/85791333/149385952-748e06d9-a952-47af-90fb-25b3d640840e.png)

**Comment by SukramJ** [comment]
As already mentioned: check with Homematic Manager at the same time. Check channel 0 on VALUES set.
We do not interpret any value. What you see is what we get from the interface. 
If the ccu displays only RSSI-Peer the just disable the RSSI-Device in HA.

**Comment by Lemocin** [comment]
@noxhirsch As you also have a `HM-ES-PMSw1-Pl-DN-R1`: Is it the same for you? 
I think if the integration gets merged into Home Assistant, we maybe have to make sure, that such unplausible values will be handled automatically if it happens for some persons. Otherwise users will surely complain.

**Comment by SukramJ** [comment]
If a device has a param in paramset then it is created. There is no value while creating.
**If the device delivers an unwanted value after initialization in HA then just disable it.** 
This is a simple autogenerated sensor, that should have no logic. It's really simple to document. Same is with Operating Voltage.
If you start to define, if a sensor should be available for a certain device, then you are on the level of pyhomematic, where every device has to be defined before usage. This is not wanted.

**Comment by noxhirsch** [comment]
@Lemocin I've just enabled the RSSI sensors for this device, but for me both RSSI sensors are available.
![image](https://user-images.githubusercontent.com/30938717/149404530-20df0791-3041-4cee-a312-b82bafafc8fb.png)
![image](https://user-images.githubusercontent.com/30938717/149404660-d17cf7a0-bcba-4315-8667-2c4164228f7e.png)

But I noticed that my HM-WDS30-OT2-SM has no "RSSI Device" and shows this instead:
![image](https://user-images.githubusercontent.com/30938717/149405614-e35f8c03-7cb0-486b-90a8-ebc91c23b160.png)
![image](https://user-images.githubusercontent.com/30938717/149405639-d00367a5-ff57-4da1-9d7f-e92c390875ce.png)

**Comment by noxhirsch** [comment]
> If a device has a param in paramset then it is created. There is no value while creating.
If the device delivers an unwanted value after initialization in HA then just disable it.

I've read you comment after posting mine. I think this is fine, since the sensors are disabled by default and I think most people wont use them anyway.

**Comment by Lemocin** [comment]
> are disabled by default

Oh okay, i didn't noticed that. I have installed the Integration at the time where all sensors were enabled. If it is disabled by default i agree, that it should be fine.

**Comment by SukramJ** [comment]
The problem is that there are flaky devices that sometimes have good values and sometimes bad values. An automatic disable is also not an option (i tried it initially), because a device could be unavailable during startup and that might be even more confusing, and it is not possible to automatically re add a device.
Good news is that this is basically a problem of the diagnostic entities, and except from low_bat these entities
are disabled by default.

**Comment by SukramJ** [comment]
Can this be closed?

**Comment by eikowagenknecht** [comment]
RSSI values can only range from -120 to 0 per definition. Also the Homematic devices store it in a datatype that can only go to -127/-128 (not exactly sure, would have to look it up).

So why not just filter out all other values and set them to None / unknown / not available? This would not be device specific.

Also there are more issues with RSSI right now:
- In the screenshot from @noxhirsch you can see that ws_charger Rssi Peer is 55 dB. This can not be. Probably it is -55 dB.
- The unit of RSSI is dBm, not dB
- For some devices the RSSI values reported from Homematic are broken unfortunately. And by broken I mean that someone forgot which type of integer they use and confuded a signed int with an unsigned one...
![image](https://user-images.githubusercontent.com/1475672/149639783-bd2c4e93-c5b2-49e4-8185-143f8ecf5d9a.png)

- This value (-199 dBm) aktually means "(Anzeigewert * -1) -256" = -57 dBm. This is a long standing issue and is discussed for example here: https://github.com/jens-maus/RaspberryMatic/issues/897 Other platforms (ioBroker) also implement a workaround for this.

So the logic would be (unless there are even more wirdnesses in what Homematic is reporting):
| Range | Value to show | Reason |
|--------------------|----------------|---------------------------------|
| < -256             | None/unknown   | Invalid                                |
| >= -256 && <= -146 | (value*-1)-256 | The "Homematic is broken" range |
| > -146 && < -120   | None/unknown   | Invalid                                |
| >= 120 && <=0      | value          | The real range                  |
| > 0 && <= 120               | value*-1    | Maybe needed if the values from @noxhirsch are reported that way from Homematic, otherwise another None/unknown range                                |
| > 120 && < 146     | None/unknown   | Invalid                                |
| >= 146 && <= 256 | value-256 | Maybe another "Homematic is broken" range, needs to be checked |
| > 256     | None/unknown   | Invalid                                |

Now that I typed it all out yes, that looks a bit insane, but it would handle all broken Homematic ranges and be robust enough that when (or rather if) Homematic finally fixes this bug it would not break.

Of course we could rightfully argue that it's Homematics job to deliver correct values, but this is known for years and they don't act on it, so maybe a workaround is the right thing here.

If None/unknown is not an option for some startup reason, I'd suggest going with a 0 instead in those cases as noone will ever have a value that good in a real use case. -65535 is apparent as a maxed out unsigned int16 (times -1) to us, but an average user might be confused by that ;-)

**Comment by danielperna84** [comment]
> Of course we could rightfully argue that it's Homematics job to deliver correct values, but this is known for years and they don't act on it, so maybe a workaround is the right thing here.

I don't know for sure, but I would assume this would require firmware updates of the devices. And if only (or at least primarily) old RF devices expose this problem, it's _better_ to not fix the issue so customers buy new devices. At least that's how I would handle it if I were eq3. 😅

**Comment by SukramJ** [comment]
@eikowagenknecht is there any evidence for your table or an existing implementation in another hm related project?
If you or someone else is willing to contribute a documentation about the RSSI related problems with HM devices and a description about your solution (table), then i could add an implementation.
A referencable documentation is mandatory, to get rid of that topic as much as possible. Hopefully others with join too.
This is especially important, because this is the first time we would manipulate received data from the CCU.

@danielperna84 is this a way we should go?

**Comment by SukramJ** [comment]
Resulting python code will 
```
def fix_rssi(value: Any) -> int | None:
    """Fix rssi value."""
    if value is None:
        return None
    if -120 <= value <= 0:
        return value
    if 0 < value <= 120:
        return value*-1
    if -256 <= value <= -146:
        return (value*-1)-256
    if 146 <= value <= 256:
        return value - 256
    return None
```

**Comment by danielperna84** [comment]
If the statements from above are true, then yes, fixing the values seems appropriate. It shouldn't be this way, but I don't expect this to be fixed by the vendor. And not fixing it this way will probably result in more support requests.

**Comment by eikowagenknecht** [comment]
> @eikowagenknecht is there any evidence for your table or an existing implementation in another hm related project?

Evidence is my own experience from what I can see in the CCU3 Web Interface and the API plus quite some posts all around the place (Raspberrymatic, FHEM, Homematic Forum, ELV Forum, ...). Some references to other sources where this problem occurs:

- https://homematic-forum.de/forum/viewtopic.php?t=58067
  - Other people coming too the comclusion that subtracting 256 is a good idea
- https://homematic-forum.de/forum/viewtopic.php?t=69105
  - > On reception, the radio CPU reads the RSSI from an underlying receiver. The RSSI is returned in result byte 2 (bit 23–16) of CMDSTA (see Figure 23-5). The RSSI is given on signed form in dBm. If no RSSI is available, this is signaled with a special value of the RSSI (−128, or 0x80). (https://www.ti.com/lit/pdf/swcu117 for technical data of the chip)
- https://forum.fhem.de/index.php?topic=106900.0
  - Values in the range of 146 to 256 described here in some circumstances, the developer of FHEM says he will implement a workaround
- https://github.com/jens-maus/RaspberryMatic/issues/897
  - Validation by Jens Maus (Raspberrymatic main dev) that values 65536, 0 and 181 can be seen and that 65536 means n/a
- https://github.com/ioBroker/ioBroker.hm-rega/commit/a8c0323917bcd6d6003668df0c74ed6ab5f8ea26
  - This is the implementation of iobroker rega adapter (though there only a default 256 seems to be subtracted which I think is not enough because we have many more weird cases, see my table)

So in fact it seems that one or two special cases seem to be missing in my table:
- 0 sometimes means "no value known", can confirm I saw this in the Homematic WebUI interface myself
- 1 apparrently can also be a false value (see opening post in https://github.com/jens-maus/RaspberryMatic/issues/897)
so the range 0 to 1 (including both endpoints) should be translated to None as well. Noone can have that good values in real life anyways.

Also I could find no reliable source of the minimal value being 120, so it is probably better to extend this to 127 since this is the maximum that Homematic devices can return because of the int type. I also fixed some edge cases (0, 256, -256). So an updated table would be:

| Range | Value to show | Reason |
|--------------------|----------------|---------------------------------|
| <= -256             | None/unknown   | Invalid                                |
| > -256 && < -129 | (value*-1)-256 | "Homematic is broken" range (translates to <0 && >-127) |
| >= -129 && <= -127   | None/unknown   | Invalid                                |
| > -127 && < 0      | value          | The real range                  |
| >= 0 && <= 1   | None/unknown   | "Homematic is broken" range (translates to "None")                               |
| > 1 && < 127               | value*-1    | "Homematic is broken" range (translates to <-1 && >-127)                               |
| >= 127 && <= 129     | None/unknown   | Invalid                                |
| > 129 && < 256 | value-256 | "Homematic is broken" range (translates to <0 && >-127) |
| >= 256     | None/unknown   | Invalid                                |

With interpreting value 1 as None we effectively loose part of the range (0 to 1) for one of the broken ranges. I think this is okay, the best RSSI value I have ever seen anywhere for Homematic is sth. like -14 dBm and this is the device directly next to the antenna, so I think it should be fine.

> If you or someone else is willing to contribute a documentation about the RSSI related problems with HM devices and a description about your solution (table), then i could add an implementation.

Sure, the table is already done and I can do some writeup when there is a place to put it :-) I think this issue has most of the information now to be able to do that without too much effort.

**Comment by SukramJ** [comment]
ok. References are good.
Could you please rephrase it to make it a wiki page?

**Comment by eikowagenknecht** [comment]
Sure, target audience is the average CCU3 user or a developer? ;-)

**Comment by SukramJ** [comment]
I think the CCU user should be the target audience.

**Comment by SukramJ** [comment]
I think best option will be to create a pull request and modify https://github.com/danielperna84/hahomematic/blob/devel/docs/rssi_fix.md

**Comment by eikowagenknecht** [comment]
Alright, I created a PR with a first draft of the documentation.

**Comment by SukramJ** [comment]
@eikowagenknecht 
Please confirm that the python code is correct:

```
    if -127 < value < 0:
        return value
    if 0 < value < 127:
        return value * -1
    if -256 < value < -129:
        return (value * -1) - 256
    if 129 < value < 256:
        return value - 256
    return None
```

**Comment by eikowagenknecht** [comment]
```
if value is None:
        return None
```
This is missing from the previous snippet.

```
if 0 < value < 127:
```
Should be 1 instead of 0 to accomodate the "1 = None" special case (see my table and the linked threads).

Looks good otherwise.

**Comment by SukramJ** [comment]
Fixed in 0.23.0
